src/main/kotlin/com/ampnet/web3provider/service/impl/ProviderServiceImpl.kt

```diff
@@ -9,12 +10,19 @@ import java.math.BigInteger
 
 @AutoJsonRpcServiceImpl
 @Service
-class ProviderServiceImpl(private val web3jService: Web3jService) : ProviderService {
+class ProviderServiceImpl(
+    private val web3jService: Web3jService,
+    private val providerRedisRepository: ProviderRedisRepository
+) : ProviderService {
 
     companion object : KLogging()
 
     override fun getBalance(address: String, blockParameter: Any): BigInteger {
         logger.info { "Received request to get eth_getBalance for address: $address and block: $blockParameter" }
-        return web3jService.getBalance(address, blockParameter)
+        val cacheBalance = providerRedisRepository.getBalance(address, blockParameter)
```

Comment:
Spring offers caching via annotations, so you wouldn't have to read the cache manually. This may be useful: https://www.baeldung.com/spring-boot-redis-cache
I'm not sure how it works with Kotlin though.

Reply:
Yeah I know it can be done via annotations, I had issues setting the cache with annotations on the service layer and 
although this way it is more verbose I would leave it as is.